identity-iota-core/src/revocation/embedded_revocation_list.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use core::fmt::Formatter;
+use core::fmt::Result as FmtResult;
+use std::io::Write;
+
+use flate2::write::ZlibDecoder;
+use flate2::write::ZlibEncoder;
+use flate2::Compression;
+use identity_core::common::Url;
+use identity_core::utils::decode_b64;
+use identity_core::utils::encode_b64;
+use roaring::RoaringBitmap;
+use serde::de;
+use serde::de::Deserializer;
+use serde::de::Visitor;
+use serde::ser::Error as _;
+use serde::ser::Serializer;
+use serde::Deserialize;
+use serde::Serialize;
+
+use super::error::Result;
+use super::error::RevocationMethodError;
+use crate::service::EmbeddedRevocationEndpoint;
+
+pub const EMBEDDED_REVOCATION_METHOD_NAME: &str = "EmbeddedRevocationList";
+const REVOCATION_LIST_INDEX: &str = "revocationListIndex";
+
+/// A compressed bitset for managing credential revocation.
+#[derive(Clone, Debug, PartialEq)]
+pub struct EmbeddedRevocationList(RoaringBitmap);
+
+impl EmbeddedRevocationList {
+  /// Returns the name of the revocation method.
+  pub fn name() -> &'static str {
+    EMBEDDED_REVOCATION_METHOD_NAME
+  }
+
+  // Returns the name of the property that contains the index of the credential to be checked.
+  pub fn credential_list_index_property() -> &'static str {
+    REVOCATION_LIST_INDEX
+  }
+
+  /// Creates a new `EmbeddedRevocationList` revocation method.
+  pub fn new() -> Self {
+    Self(RoaringBitmap::new())
+  }
+
+  /// Returns `true` if the credential at the given `index` is revoked.
+  pub fn is_revoked(&self, index: u32) -> bool {
+    self.0.contains(index)
+  }
+
+  /// Revokes the credential at the given `index`.
+  pub fn revoke(&mut self, index: u32) -> bool {
+    self.0.insert(index)
+  }
+
+  /// Given the index of multiple credentials, revoke all.
+  pub fn revoke_multiple(&mut self, indexes: &[u32]) {
+    for i in indexes {
+      self.revoke(*i);
+    }
+  }
+
+  /// The credential at the given `index` will be set to valid.
+  pub fn undo_revocation(&mut self, index: u32) -> bool {
+    self.0.remove(index)
+  }
+
+  /// Serializes and compressess the [`EmbeddedRevocationList`] and returns its data url representation
+  pub fn to_embedded_service_endpoint(&self) -> Result<EmbeddedRevocationEndpoint> {
+    let data_url: String = format!("data:,{}", self.serialize_compressed_b64()?);
+    EmbeddedRevocationEndpoint::parse(&data_url).map_err(|_| RevocationMethodError::InvalidUrlRepresentation(data_url))
+  }
```

Comment:
Nit: this should probably be a `TryFrom` implementation and specified in `EmbeddedRevocationEndpoint`.